### PR TITLE
fix(melange): run melc ppx with merlin

### LINKED ISSUE #6456: bug(melange): merlin does not know about pipe first (`|.`) operator
When running latest version of `dune` from `main` in a Melange project, usages of fast pipe operator can't be processed by merlin, which returns this error for each appearance:

```
Unbound value |.
```

After discussing with @anmonteiro, it seems merlin conf generated by Dune should include something like the following:
```
FLG -ppx 'melc -as-ppx -bs-jsx 3'
```

### PR DESCRIPTION
**Pull request by jchavarri**
Fixes #6456.

A couple of questions:
- I am not sure what to do with `to_dot_merlin`. It seems all `FLG` directives are commented? I don't understand what's the purpose
- Unsure if `-bs-jsx 3` should be conditionally added, based on `flags` field value on `library` or `melange.emit` stanzas.

### CODE REVIEW

File: src/dune_rules/merlin.ml [hunk 1]
```diff
@@ -94,11 +95,20 @@ module Processed = struct
...
+            (Sexp.List [ Atom (Pp_kind.to_flag flag); Atom args ])
+          :: flags
+      in
+      match mode with
+      | `Ocaml -> flags
+      | `Melange ->
         make_directive "FLG"
-          (Sexp.List [ Atom (Pp_kind.to_flag flag); Atom args ])
+          (Sexp.List
+             [ Atom (Pp_kind.to_flag Ppx); Atom "melc -as-ppx -bs-jsx 3" ])
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:111]: Shouldn't the melange ppx come first?

File: src/dune_rules/merlin.ml [hunk 2]
```diff
@@ -414,7 +435,7 @@ module Unprocessed = struct
         Path.Set.union src_dirs
           (Path.Set.of_list_map ~f:Path.source more_src_dirs)
       in
-      { Processed.stdlib_dir; src_dirs; obj_dirs; flags; extensions }
+      { Processed.stdlib_dir; src_dirs; obj_dirs; flags; extensions; mode }
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:438]: Processed shouldn't include the mode. The flags should already contain what's needed.

File: src/dune_rules/merlin.ml (hunk 1 above)
Comment (jchavarri) [reply src/dune_rules/merlin.ml:111]: Isn't the merlin syntax something like `FLG -ppx melc -as-ppx -bs-jsx 3`? The first atom is what renders the `-ppx` flag next to `FLG`.

File: src/dune_rules/merlin.ml (hunk 2 above)
Comment (jchavarri) [reply src/dune_rules/merlin.ml:441]: Sorry, I spent some time trying to implement this, but I don't understand how it is possible.

`flags` includes only the compiler flags, and it's dumped in a single `FLG` directive here:

https://github.com/ocaml/dune/blob/75d5b4083857bee66f69b217aa6ca94147c2475a/src/dune_rules/merlin.ml#L92-L95

We want melange `FLG` to be a separate directive. I can replace `mode` field in processed with something like `melange_ppx_flag` of type, but something else needs to be passed to processed in order to add this new directive.

**Comment by jchavarri** [comment]
> It should be always added for emit stanzas and for libraries that are "melange only". See the condition we have for object directories already.

I meant something else, rather how to pass compiler flags to melange `-ppx` directive. I managed to do it in https://github.com/ocaml/dune/pull/6476/commits/5de07976baefcbd1879a67b6f900f65727f2670f cc @anmonteiro

**Comment by jchavarri** [comment]
> I meant something else, rather how to pass compiler flags to melange -ppx directive. I managed to do it in https://github.com/ocaml/dune/commit/5de07976baefcbd1879a67b6f900f65727f2670f

I tested this change with vscode editor and it seems it doesn't work, I started getting errors of the kind `unknown flag -as-ppx`. It seems they come from merlin?

I am not sure how to fix this, so I reverted the change in https://github.com/ocaml/dune/pull/6476/commits/993c0ffbb5f8a1e0b9f5dacd9867d712df9ed965.

**Comment by anmonteiro** [comment]
- We don't need to pass all the compiler flags to `melc --as-ppx`, just the `-bs-jsx 3` flag
- It's not the end of the world to pass `-bs-jsx 3` unconditionally, it just means that the PPX will always be applied
  - I highlighted a plan to improve this in https://github.com/ocaml/dune/pull/6476#discussion_r1024322957

File: src/dune_rules/merlin.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_rules/merlin.ml:438]: Why does it need to be a separate `FLG` directive? We manage to put all the other flags in one directive.

Comment (jchavarri) [reply src/dune_rules/merlin.ml:442]: @rgrinberg Any thoughts on the above? Considering there were already two `FLG` directives, should `melc` ppx be made part of any of them or could it exist on its own? If anything, it should  be part of the second, but I don't think it's possible at the moment as it's not a user defined ppx (so can't be compiled with ppx driver).

Comment (rgrinberg) [reply src/dune_rules/merlin.ml:442]: This is hacky but it looks good enough for now. A few points:

1. We shouldn't hard code the path of `melc`. We should use the absolute path that was resolved by dune
2. We should make sure that `melc` is actually built if it's defined in the same workspace.

In any case, if @anmonteiro is fine with this temporary hack then we can merge it.

File: test/blackbox-tests/test-cases/melange/merlin.t [hunk 3]
```diff
@@ -10,6 +10,7 @@
   > (library
   >  (name $lib)
   >  (private_modules bar)
+  >  (flags :standard -bs-jsx 3)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/merlin.t:13]: why do we set this flag manually? isn't this already hard coded by this PR?

File: test/blackbox-tests/test-cases/melange/merlin.t [hunk 4]
```diff
@@ -37,3 +45,8 @@
   $ dune build @check
   $ dune ocaml-merlin --dump-config="$PWD" | grep -i "$target"
     $TESTCASE_ROOT/_build/default/.output.mobjs/melange)
+
+The melange.emit entry contains a ppx directive
+
+  $ dune ocaml-merlin --dump-config="$(pwd)" | grep -i "ppx"
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/merlin.t:51]: I change the command to `dune ocaml merlin dump-config $PWD` in the main branch

File: src/dune_rules/merlin.ml (hunk 2 above)
Comment (anmonteiro) [reply src/dune_rules/merlin.ml:442]: I agree we should fix 1. & 2. before merging.

What's the temporary hack? the 2 `FLG` directives?

Comment (rgrinberg) [reply src/dune_rules/merlin.ml:442]: Yes. I would expect all these melange ppx's to be ported to ppxlib and enabled explicitly.

Comment (anmonteiro) [reply src/dune_rules/merlin.ml:442]: that's the goal for the ReactJS PPX, as highlighted in https://github.com/ocaml/dune/pull/6476#discussion_r1024322957.

I don't yet have a clear plan for the builtin PPX, which currently is very tightly integrated into the compiler.

File: test/blackbox-tests/test-cases/melange/merlin.t (hunk 3 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/merlin.t]: it was used for previous attempts to connect flags from `dune` file into merlin directive. Removed in https://github.com/ocaml/dune/pull/6476/commits/858dac6666a97ce0b573c5dd8b6499175efd3480.

File: test/blackbox-tests/test-cases/melange/merlin.t (hunk 4 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/merlin.t]: https://github.com/ocaml/dune/pull/6476/commits/4acd9e1e11639e22b808239e38d1e92108146147

**Review by anmonteiro — APPROVED** [review]
Looks good. Do you plan to fix the remaining TODO in `Merlin.to_dot_merlin`?

**Comment by jchavarri** [comment]
> Do you plan to fix the remaining TODO in `Merlin.to_dot_merlin`?

Fixed the todo and verified the output in test in 
https://github.com/ocaml/dune/pull/6476/commits/c4a6d2bb8cf844268f9b930b746387b8f64f3a18.

File: test/blackbox-tests/test-cases/melange/merlin.t [hunk 5]
```diff
@@ -35,5 +42,18 @@
...
+  $ dune ocaml merlin dump-config $PWD | grep -i "ppx"
+   (FLG (-ppx "melc -as-ppx -bs-jsx 3"))
+
+  $ dune ocaml dump-dot-merlin $PWD
+  EXCLUDE_QUERY_DIR
+  STDLIB /home/me/code/dune/_opam/lib/melange
+  B $TESTCASE_ROOT/_build/default/.output.mobjs/melange
+  S $TESTCASE_ROOT
+  # FLG -w @1..3@5..28@30..39@43@46..47@49..57@61..62@67@69-40 -strict-sequence -strict-formats -short-paths -keep-locs
+  # FLG -ppx melc -as-ppx -bs-jsx 3
```
Comment (anmonteiro) [inline test/blackbox-tests/test-cases/melange/merlin.t:59]: I think this needs to be `FLG -ppx 'melc -as-ppx -bs-jsx 3'`

File: src/dune_rules/merlin.ml [hunk 6]
```diff
@@ -145,6 +156,11 @@ module Processed = struct
           print "# FLG";
           List.iter flags ~f:(fun f -> printf " %s" (quote_for_dot_merlin f));
           print "\n");
+    let () =
+      match mode with
+      | `Ocaml -> ()
+      | `Melange -> print "# FLG -ppx melc -as-ppx -bs-jsx 3\n"
```
Comment (anmonteiro) [inline src/dune_rules/merlin.ml:162]: I think we can use the same machinery above, with `quote_for_dot_merlin`. See my comment in the test.

Comment (jchavarri) [reply src/dune_rules/merlin.ml]: Please check https://github.com/ocaml/dune/pull/6476/commits/5c34973f817c0dab8da113fcb6849b3898eb0520.

File: test/blackbox-tests/test-cases/melange/merlin.t (hunk 5 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/merlin.t]: Fixed in https://github.com/ocaml/dune/pull/6476/commits/5c34973f817c0dab8da113fcb6849b3898eb0520.

**Review by anmonteiro — APPROVED** [review]

**Comment by rgrinberg** [comment]
`melc` is still looked up in `$PATH` and there's no dependency on it

**Comment by jchavarri** [comment]
@rgrinberg Could you elaborate please? You mean `melc` should be specified as dep for the tests, for the build or for Merlin itself?

For the tests, we add it here
https://github.com/ocaml/dune/blob/dd6f9763cde3ecb30f2af4adc59f92eedb0de8ae/test/blackbox-tests/test-cases/melange/dune#L2

For the build rules, we don't add dep explicitly, but check and error out if the executable is not found (what we did in https://github.com/ocaml/dune/pull/6403/).

For Merlin, I am not sure how this should work. Merlin might rely on a few binaries to exist (ocaml toolchain, user ppxs, etc) but I can't find an example that shows how one can define a dependency on binaries. I can only see directives `B`, `S`, `FLG`, `STDLIB`, `EXCLUDE_QUERY_DIR`, but idu how any of them could be used to define a dependency on a binary.

**Comment by rgrinberg** [comment]
>  Merlin might rely on a few binaries to exist (ocaml toolchain, user ppxs, etc) but I can't find an example that shows how one can define a dependency on binaries. I can only see directives B, S, FLG, STDLIB, EXCLUDE_QUERY_DIR, but idu how any of them could be used to define a dependency on a binary.

Note that user ppx's are provided as absolute paths to merlin for exactly this reason. Merlin doesn't use the ocaml toolchain (it vendors what it needs), so there's no need to worry about that.

Okay, we can skip the dependency on melc for now because we don't handle ppx cleanly either.

File: src/dune_rules/merlin.ml [hunk 7]
```diff
@@ -68,7 +69,9 @@ module Processed = struct
 
   let serialize_path = Path.to_absolute_filename
 
-  let to_sexp ~pp { stdlib_dir; obj_dirs; src_dirs; flags; extensions } =
+  let melc_ppx_flg = "melc -as-ppx -bs-jsx 3"
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:72]: This is the offending line. It should be using `serialize_path` to encode `melc` and using the `melc` path we resolved to build the actual library

Comment (jchavarri) [reply src/dune_rules/merlin.ml:72]: > It should be using `serialize_path` to encode `melc` and using the `melc` path we resolved to build the actual library

@rgrinberg Could you help me understand this part, please? The path that is resolved in merlin points to the melc stdlib e.g. `_opam/lib/melange`, assuming you refer to this code

https://github.com/ocaml/dune/blob/5a098272172c1f16aad309d6aabc016f7be2a9f3/src/dune_rules/merlin.ml#L398

But the path here should be to the `melc` command, which would be in e.g. `_opam/bin/melc`. How are both related? Or am I misunderstanding?

Comment (rgrinberg) [reply src/dune_rules/merlin.ml:72]: Sure, keep asking questions if something is unclear. The gist of the issue is: we should not use any relative paths in the .Merlin directive. All paths should be absolute. So, library paths, stdlib paths, preprocessors, etc should all be fully resolved.

The current code violates this with the way it refers to the melc binary.

Comment (jchavarri) [reply src/dune_rules/merlin.ml]: Thanks for clarifying. I tried a fix in https://github.com/ocaml/dune/pull/6476/commits/c6e8bd64566396880daf728501160d9aa6fa677b.

**Comment by jchavarri** [comment]
@rgrinberg in https://github.com/ocaml/dune/pull/6476/commits/f698bd659ec5ece3141ed61e7d79741c344bd5b0 I lifted the calculation of melc flags to the `Unprocessed.process` function. This allows to share the flags between both `to_dot_merlin` and `to_sexp` functions. I think the code is cleaner now, but please let me know your thoughts.

File: src/dune_rules/merlin.ml [hunk 8]
```diff
@@ -183,32 +202,42 @@ module Processed = struct
...
              ->
             ( pp_config :: acc_pp
             , Path.Set.union acc_obj obj_dirs
             , Path.Set.union acc_src src_dirs
             , flags :: acc_flags
-            , extensions @ acc_ext ))
+            , extensions @ acc_ext
+            , match acc_melc_flags with
+              | Some _ -> acc_melc_flags
+              | None -> melc_flags ))
```
Comment (jchavarri) [inline src/dune_rules/merlin.ml:236]: I am not sure about this. Should it rather be `melc_flags` directly? (i.e. no accumulation needed)

I guess the missing part is: what is the relationship between the `paths` passed as param to `print_generic_dot_merlin`?

File: src/dune_rules/merlin.ml [hunk 9]
```diff
@@ -120,7 +130,8 @@ module Processed = struct
     in
     if String.need_quoting s then Filename.quote s else s
 
-  let to_dot_merlin stdlib_dir pp_configs flags obj_dirs src_dirs extensions =
+  let to_dot_merlin stdlib_dir pp_configs flags obj_dirs src_dirs extensions
+      melc_flags =
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:134]: Do we need to pass these melange flags separately? They can be just a part of `flags`.

Comment (jchavarri) [reply src/dune_rules/merlin.ml:134]: It is not possible, or I don't see how. We talked about this earlier, please check https://github.com/ocaml/dune/pull/6476#discussion_r1026292933 for details.

Comment (rgrinberg) [reply src/dune_rules/merlin.ml]: I pushed a commit describing what I meant.

File: src/dune_rules/merlin.ml [hunk 10]
```diff
@@ -33,6 +33,7 @@ module Processed = struct
     ; src_dirs : Path.Set.t
     ; flags : string list
     ; extensions : string Ml_kind.Dict.t list
+    ; melc_flags : string list option
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:36]: What's the difference between `None` and `Some []`?

File: src/dune_rules/merlin.ml (hunk 9 above)
Comment (jchavarri) [reply src/dune_rules/merlin.ml]: I see, thanks, I missed that `flags` in this function has some different semantics. Updated test in https://github.com/ocaml/dune/pull/6476/commits/437fabd9ae8254af1b62cdec23b6d091c90d6c94.

File: src/dune_rules/merlin.ml (hunk 10 above)
Comment (jchavarri) [reply src/dune_rules/merlin.ml:36]: `None` is used when we can't find the `melc_compiler`. I understand we don't want to fail at the merlin stage, like in other cases, e.g.

https://github.com/ocaml/dune/blob/437fabd9ae8254af1b62cdec23b6d091c90d6c94/src/dune_rules/merlin.ml#L345

Comment (rgrinberg) [reply src/dune_rules/merlin.ml:36]: Yes, but then why not return `[]`?

Comment (jchavarri) [reply src/dune_rules/merlin.ml]: I guess we can, I thought `option` was more explicit. Updated in https://github.com/ocaml/dune/pull/6476/commits/dfff493f13a12a6802f826f4a9e56fde49385d5c.

**Comment by rgrinberg** [comment]
According to the tests, `melc` is being added even when it's not being used.

**Comment by jchavarri** [comment]
@rgrinberg Fixed in https://github.com/ocaml/dune/pull/6476/commits/dadc8962e9560cde6a1bc49f01cce76b15cd7fe4. I moved the definition of `melc_compiler` inside the code block where `melc_flags` is calculated in order to follow the style guidelines about scope.

**Comment by jchavarri** [comment]
I have been testing vscode-ocaml extension using this PR, but unfortunately something regressed with the most recent changes and now it does not work at all.

I see errors in lsp log of the kind:
```
melc: can not handle multiple files
unknown flag 3
unknown flag -as-ppx
```

This used to work at the beginning of the PR, when passing whole command in quotes, e.g. `(FLG (-ppx "melc -as-ppx -bs-jsx 3"))`. Is there a way to do something similar with sexp atoms?

**Comment by rgrinberg** [comment]
> This used to work at the beginning of the PR, when passing whole command in quotes, e.g. (FLG (-ppx "melc -as-ppx -bs-jsx 3")). Is there a way to do something similar with sexp atoms?

Not that I know. Let's just leave it as a single form for now.

**Pull request merged**